src/recursive_shared_mutex.h

```diff
@@ -0,0 +1,101 @@
+#ifndef _RECURSIVE_SHARED_MUTEX_H
+#define _RECURSIVE_SHARED_MUTEX_H
+
+#include <cassert>
+#include <chrono>
+#include <condition_variable>
+#include <exception>
+#include <map>
+#include <mutex>
+#include <system_error>
+#include <thread>
+#include <tuple>
+#include <type_traits>
+
+
+/*
+ * This mutex has two levels of access, shared and exclusive. Multiple threads can own this mutex in shared mode but
+ * only
+ * one can own it in exclusive mode.
+ * - A thread is considered to have ownership when it successfully calls either lock or try_lock.
+ * - A thread may recusively call lock for ownership and must call a matching number of unlock calls to end ownership.
+ * - A thread MAY call for shared ownership if it already has exclusive ownership. This should just add an additional
+ * lock on
+ * top of write counter and not actually lock. If a thread has exclusive ownership and checks for shared ownership this
+ * should return true.
+ * - A thread MAY obtain exclusive ownership if no threads excluding itself has shared ownership. (this might need to
+ * check
+ * for another write lock already
+ * queued up so we dont jump the line)
+ */
+
+
+static const std::thread::id NON_THREAD_ID = std::thread::id();
+
+class recursive_shared_mutex
+{
+private:
```

Comment:
Please make variables protected not private so that test and debugging code can derive from your class and break through the veil.

Reply:
good idea, will do